### ISSUE TITLE
For MSUs - check if Windows Update is disable and error out + Auto enable switch?

### ISSUE THREAD

**Issue by DarwinJS** [chocolatey/choco#1073]
It is a common and good practice to disable windows update for long running software deployment orchestration so that it does not clash with the activities of the orchestration code.  However, some Chocolatey packages have to run MSUs.

Would be nice if Choco explicitly checked if windows update is in a disabled state when it is about to install an MSU.

Would be helpful if the error message suggested that windows update be enabled before the package run and disable again after.

Would be AWESOME if there was a switch something like "--auto-enable-WindowsUpdate-when-needed"

Then it could be enabled at the last possible moment before Chocolatey needs it and disabled at the first possible moment after - this would reduce the possibility that Windows update takes off on an update cycle before orchestration code calling the package could disable it again.

Using the switch would still carry this risk - it might be helpful to explain that in the switch help.

**Comment by jberezanski** [comment]
Did you have an actual case when an MSU failed to install with Windows Update disabled?

As a general rule, I have Automatic Updates disabled ([IAutomaticUpdatesSettings::NotificationLevel](https://msdn.microsoft.com/en-us/library/windows/desktop/aa385824(v=vs.85).aspx) [1=aunlDisabled](https://msdn.microsoft.com/en-us/library/windows/desktop/aa385806(v=vs.85).aspx)) on my test machines, to maintain a consistent and repeatable testing environment, and MSUs install just fine.

**Comment by jberezanski** [comment]
Unless you're talking about turning off and disabling the Windows Update service (which might be the only way, albeit an unsupported one, to pause updates on Windows 10).

**Comment by DarwinJS** [comment]
Yes I am talking about disabling the service.  Boxstarter has a built-in standard command for this.  

IMHO (and the practice of many others) the implications of a concurrently running wIndows update cycle running on an older boot of windows (that is typical with ISO builds) far outweigh whether disabling Windows Update service is "supported" :)

**Comment by jberezanski** [comment]
Fair point. In my orchestrations of Windows 10 I usually start with installing all available updates, to minimize the chance of interference from Windows Update. (Although I found the need to disable other auto updaters, such as Adobe, as well.)

In that case, yes, a clear error message from choco would be useful.

I'm not sure about the switch, though. Extensive testing would need to be performed to determine if the service could be disabled immediately after installing the MSU, or if it would need to remain active during the subsequent reboot.
Depending on the internal workings of the WU service, it could even decide to install pending updates first, and only then proceed to install the provided MSU.

I'm toying with another idea now. MSU files are ZIP archives, which usually contain installation packages in the form of CAB files. Those CAB files are installable using dism.exe or the Add-WindowsPackage cmdlet. Perhaps that way we could bypass the WU service entirely, at least for MSUs which do not contain any additional logic or executables...

**Comment by jberezanski** [comment]
I used this approach recently when deploying RSAT over PowerShell Remoting. wusa.exe does not work through Remoting (same as the Windows Update Agent API), so I extracted cabs out of it and applied them using dism.

It is even documented by Microsoft, in a fashion: [KB article](https://support.microsoft.com/en-us/kb/934307) [blog post](https://blogs.msdn.microsoft.com/astebner/2008/03/11/knowledge-base-article-describing-how-to-extract-msu-files-and-automate-installing-them/)

**Comment by jberezanski** [comment]
Of course, it would be feasible only if installing cabs using dism or Add-WindowsPackage did not require wuauserv to be running - needs testing.

**Comment by dragon788** [comment]
I also ran across a bug on AppVeyor recently that seemed to have this same problem, as apparently in that environment the wusa service is disabled, so it couldn't apply an .msu which failed the user's build since they were trying to install a chocolatey package as a prerequisite to something else.

**Comment by dragon788** [comment]
See also this excellent example of enabling the Windows Update service to install a package that has a dependency on a KB that requires wusa be active. https://github.com/richardschneider/net-ipfs-mount/blob/be3a73e9b72bd67156a842d0099817c299ce1bbc/appveyor.yml

**Comment by pauby** [comment]
Thanks for raising this issue. After reviewing it, it's not something we expect to prioritise in the foreseeable future. We’re closing it to keep our backlog focused, but we’re open to revisiting it in the future if priorities shift or new interest arises. You are welcome to comment here or open a new issue if you'd like to discuss it further down the line.
